automatilib/cola_auth/i_dot_ai_cola_views.py

```diff
@@ -0,0 +1,155 @@
+    cola_issuer = f"https://cognito-idp.{settings.AWS_REGION_NAME}.amazonaws.com/{settings.COLA_COGNITO_USER_POOL_ID}"
+    response = requests.get(f"{cola_issuer}/.well-known/jwks.json", timeout=5)
+    if response.status_code != 200:
+        return None
+    cola_cognito_user_pool_jwk = response.json()
+    return cola_cognito_user_pool_jwk, cola_issuer
+
+
+class IAIColaLogout(View):
+    @abstractmethod
+    def post_logout(self):
+        """
+        A method that is invoked post logout of a user
+        """
+        pass
+
+    @abstractmethod
+    def pre_logout(self):
+        """
+        A method that is invoked pre logout of a user
+        """
+        pass
+
+    def get(self, request: HttpRequest, **kwargs: dict) -> HttpResponse:
+        """
+        Logs a user out of the system and removes the COLA JWT from their browser cookies
+        :param request: The HTTP request
+        :return: A HTTP response without the JWT token cookie
+        """
+        logout(request)
+        redirect_url = settings.LOGIN_URL
+        response = redirect(reverse(redirect_url or "index"))
+        response.delete_cookie(settings.COLA_COOKIE_NAME)
+        return response
+
+
+class IAIColaLogin(View):
+    @abstractmethod
+    def pre_login(self):
+        """
+        A method that is invoked after checking if a user is already logged in,
+        but before performing any actions on the request
+        """
+        pass
+
+    @abstractmethod
+    def post_login(self):
+        """
+        A method that is invoked after `handle_claims` and logging/authenticating a user,
+        but before returning an HTTP response
+        """
+        pass
+
+    @abstractmethod
+    def handle_user_jwt_details(self, user: User, token_payload: dict) -> None:
+        """
+        A method that is invoked after logging/authenticating a user and before `post_login`,
+        but before returning an HTTP response.
+        Email is already taken from the token and saved to the user object before this
+        :param user: The user that is logged in and authenticated
+        :param token_payload: The token payload that includes information from COLA
+        """
+        pass
+
+    def get(self, request: HttpRequest, **kwargs: dict) -> HttpResponse:
+        redirect_url = settings.LOGIN_REDIRECT_URL
+        if request.user.id:
+            return redirect(reverse(redirect_url or "index"))
+
+        self.pre_login()
+
+        if not (cola_cookie := request.COOKIES.get(settings.COLA_COOKIE_NAME, None)):
+            LOGGER.error("No cookie found")
+            return HttpResponseServerError()  # TODO: Improve error handling throughout
+        if not (
+            regex_match := re.search(
+                settings.COLA_JWT_EXTRACTION_REGEX_PATTERN,
+                unquote(cola_cookie),
+            )
+        ):
+            LOGGER.error("No cookie regex match found")
+            return HttpResponseServerError()
```

Comment:
I agree that in principle this is a server error.... but I think we should treat it as an authentication error e.g. 401?